pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// reconciler implements the reconcile.Reconcile interface for ManagedSeed reconciliation.
+type reconciler struct {
+	ctx         context.Context
+	clientMap   clientmap.ClientMap
+	config      *config.GardenletConfiguration
+	imageVector imagevector.ImageVector
+	recorder    record.EventRecorder
+	logger      *logrus.Logger
+}
+
+// newReconciler returns the new ManagedSeed reconciler.
+func newReconciler(ctx context.Context, clientMap clientmap.ClientMap, config *config.GardenletConfiguration, imageVector imagevector.ImageVector, recorder record.EventRecorder, logger *logrus.Logger) reconcile.Reconciler {
+	return &reconciler{
+		ctx:         ctx,
+		clientMap:   clientMap,
+		config:      config,
+		imageVector: imageVector,
+		recorder:    recorder,
+		logger:      logger,
+	}
+}
+
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	gardenClient, err := r.clientMap.GetClient(r.ctx, keys.ForGarden())
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get garden client: %w", err)
+	}
+
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := gardenClient.Client().Get(r.ctx, request.NamespacedName, managedSeed); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Debugf("Skipping ManagedSeed %s because it has been deleted", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Errorf("Could not get ManagedSeed %s from store: %v", request.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	if managedSeed.Spec.Shoot == nil || managedSeed.Spec.Shoot.Name == "" {
+		r.logger.Errorf("Skipping ManagedSeed %s because it does not specify a shoot", request.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	if managedSeed.DeletionTimestamp != nil {
+		return r.delete(gardenClient, managedSeed)
+	}
+	return r.createOrUpdate(gardenClient, managedSeed)
+}
+
+func (r *reconciler) createOrUpdate(gardenClient kubernetes.Interface, managedSeed *seedmanagementv1alpha1.ManagedSeed) (reconcile.Result, error) {
+	managedSeedLogger := logger.NewFieldLogger(r.logger, "managedSeed", kutil.ObjectName(managedSeed))
+
+	// Ensure gardener finalizer
+	if err := controllerutils.EnsureFinalizer(r.ctx, gardenClient.DirectClient(), managedSeed, gardencorev1beta1.GardenerName); err != nil {
```

Comment:
We want to rely even more on patch requests from now on (which are also safe for use with the normal client).
```suggestion
	if err := controllerutils.PatchFinalizers(r.ctx, gardenClient.Client(), managedSeed, gardencorev1beta1.GardenerName); err != nil {
```

Reply:
OK, will take note.